pkg/roll/farm.go

```diff
@@ -0,0 +1,203 @@
+	}
+}
+
+// Start is a blocking function that monitors Consul for updates. The Farm will
+// attempt to claim updates as they appear and, if successful, will start
+// goroutines for those updatesto do their job. Closing the quit channel will
+// cause this function to return, releasing all locks it holds.
+//
+// Start is not safe for concurrent execution. Do not execute multiple
+// concurrent instances of Start.
+func (rlf *Farm) Start(quit <-chan struct{}) {
+	subQuit := make(chan struct{})
+	defer close(subQuit)
+	rlWatch, rlErr := rlf.rls.Watch(subQuit)
+
+START_LOOP:
+	for {
+		select {
+		case <-quit:
+			rlf.logger.NoFields().Infoln("Halt requested, releasing updates")
+			rlf.releaseChildren()
+			return
+		case session := <-rlf.sessions:
+			if session == "" {
+				// our session has expired, we must assume our locked children
+				// have all been released and that someone else may have
+				// claimed them by now
+				rlf.logger.NoFields().Errorln("Session expired, releasing updates")
+				rlf.lock = nil
+				rlf.releaseChildren()
+			} else {
+				// a new session has been acquired - only happens after an
+				// expiration message, so len(children)==0
+				rlf.logger.WithField("session", session).Infoln("Acquired new session")
+				lock := rlf.kps.NewUnmanagedLock(session, "")
+				rlf.lock = &lock
+				// TODO: restart the watch so that you get updates right away?
+			}
+		case err := <-rlErr:
+			rlf.logger.WithError(err).Errorln("Could not read consul updates")
+		case rlFields := <-rlWatch:
+			rlf.logger.WithField("n", len(rlFields)).Debugln("Received update update")
+			if rlf.lock == nil {
+				// we can't claim new nodes because our session is invalidated.
+				// raise an error and ignore this update
+				rlf.logger.NoFields().Warnln("Received update update, but do not have session to acquire locks")
+				continue
+			}
+
+			// track which children were found in the returned set
+			foundChildren := make(map[fields.ID]struct{})
+			for _, rlField := range rlFields {
+				rlLogger := rlf.logger.SubLogger(logrus.Fields{
+					"ru_id": rlField.NewRC,
+				})
+				if _, ok := rlf.children[rlField.NewRC]; ok {
+					// this one is already ours, skip
+					rlLogger.NoFields().Debugln("Got update already owned by self")
+					foundChildren[rlField.NewRC] = struct{}{}
+					continue
+				}
+
+				err := rlf.lock.Lock(kp.LockPath(kp.RollPath(rlField.NewRC.String())))
+				if _, ok := err.(kp.AlreadyLockedError); ok {
+					// someone else must have gotten it first - log and move to
+					// the next one
+					rlLogger.NoFields().Debugln("Lock on update was denied")
+					continue
+				} else if err != nil {
+					rlLogger.NoFields().Errorln("Got error while locking update - session may be expired")
+					// stop processing this update and go back to the select
+					// chances are this error is a network problem or session
+					// expiry, and all the others in this update would also fail
+					continue START_LOOP
```

Comment:
So, lets assume the session manager is imperfect and sends multiple session IDs without sending the `""` expiry message. Both the rc farm and the roll farm seem like they'll get stuck forever trying to use the same locks. Should we explicitly release the lock for children if we enter this error case?